### PR TITLE
Dphan/cleanup features

### PR DESCRIPTION
**Pull request by dorotaphanSiili**
- extract monitor feature and remove other uses of agent
- extend examples with monitor (if feature is on)
- clean up features
- api - adds main cmake and module cmake, api files, qml wrapper and factory files, still requires apigear feature, cause it uses simulation
- stubs (used to be lib)- add implementation, also folder lib changed to implementation
- apigear - internal feature
- examples - all examples, with olink which is used for qml exmple
- monitor- new feautre:
         -  contains agent (extended with signal tracing)
         -  InterfaceTraced class which is wrapper for Interface (obtained as a param)
          - factory wrapper, that uses injected facotry, and wraps created object with traces
- plugin renamed to qmlplugin
- other features without changes (http, olink)
- cleaned up cmakes - added project name, required cmkae version and cpp standard

### CODE REVIEW

File: goldenmaster/examples/olinkserver/main.cpp [hunk 1]
```diff
@@ -1,38 +1,55 @@
 
-#include "testbed2/lib/manyparaminterface.h"
+#include "testbed2/implementation/manyparaminterface.h"
```
Comment (jryannel) [inline goldenmaster/examples/olinkserver/main.cpp:2]: one the one hand we use lib as abbreviation for library but do not use impl for implementation. I would prefer impl over implementation.

File: goldenmaster/examples/olinkserver/main.cpp [hunk 2]
```diff
@@ -46,56 +63,56 @@ int main(){
     ApiGear::ObjectLink::RemoteRegistry registry;
     ApiGear::ObjectLink::OLinkHost server(registry);
     server.listen("localhost", 8182);
-    testbed2::ManyParamInterface testbed2ManyParamInterface;
-    auto testbed2OlinkManyParamInterfaceService = std::make_shared<testbed2::OLinkManyParamInterfaceAdapter>(registry, &testbed2ManyParamInterface);
+    auto testbed2ManyParamInterface = std::make_shared<testbed2::ManyParamInterface>();testbed2::ManyParamInterfaceTraced testbed2ManyParamInterfaceTraced(testbed2ManyParamInterface );
```
Comment (jryannel) [inline goldenmaster/examples/olinkserver/main.cpp:66]: are these on the same line, or does it looks just so in the review view.

File: templates/http/httpfactory.h.tpl [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #include <QtCore>
 #include <QtNetwork>
 
-#include "{{snake .Module.Name}}/api/apifactory.h"
+#include "{{snake .Module.Name}}/api/apifactoryinterface.h"
```
Comment (jryannel) [inline templates/http/httpfactory.h.tpl:6]: wasn't another class named "I<classname>.h" why this "...interface.h"

File: templates/examples/qml/main.cpp.tpl [hunk 4]
```diff
@@ -1,12 +1,16 @@
+{{- $features := .Features }}
 {{- range .System.Modules }}
 {{- $module := . }}
 {{- range $module.Interfaces }}
 {{- $interface := . }}
-#include "{{snake $module.Name}}/lib/{{ lower ( camel $interface.Name) }}.h"
-#include "{{snake $module.Name}}/lib/qml{{lower ( camel $interface.Name)}}.h"
+#include "{{snake $module.Name}}/implementation/{{ lower ( camel $interface.Name) }}.h"
```
Comment (jryannel) [inline templates/examples/qml/main.cpp.tpl:6]: the lib is called "<name>_impl" but the folder is called implementation. Where for a lib both names are the same.

File: templates/examples/olinkclient/main.cpp.tpl [hunk 5]
```diff
@@ -3,12 +3,15 @@
 #include "apigear/olink/olinkclient.h"
 #include "olink/clientregistry.h"
 #include <memory>
-
+{{- $features := .Features }}
```
Comment (jryannel) [inline templates/examples/olinkclient/main.cpp.tpl:6]: You can always access the root context with `$` so `$.Features` should always work. No need to change.

**Review by jryannel — CHANGES_REQUESTED** [review]
- the interface for client node was called "iclientnode.h". Should we use `I` as a prefix everywhere for interfaces?
- Due to the golden master changes I find its hard to review. Not sure how to make it better.

File: templates/http/httpfactory.h.tpl (hunk 3 above)
Comment (dorotaphanSiili) [reply templates/http/httpfactory.h.tpl:6]: you mean I should rename it to iapifactory and a class to IApiFactory?

File: goldenmaster/examples/olinkserver/main.cpp (hunk 2 above)
Comment (dorotaphanSiili) [reply goldenmaster/examples/olinkserver/main.cpp:66]: fixing

File: templates/examples/olinkclient/main.cpp.tpl (hunk 5 above)
Comment (dorotaphanSiili) [reply templates/examples/olinkclient/main.cpp.tpl:6]: thx

**Review by jryannel — APPROVED** [review]

**Pull request merged**